artifact/v3/version/version.cpp

```
@@ -0,0 +1,106 @@
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <string>
+#include <vector>
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/error.hpp>
+#include <common/json.hpp>
+
+#include <artifact/v3/version/version.hpp>
+
+namespace artifact {
+namespace v3 {
+namespace version {
+
+namespace io = mender::common::io;
+namespace expected = mender::common::expected;
+namespace error = mender::common::error;
+namespace json = mender::common::json;
+
+const int supported_parser_version {3};
+const string supported_parser_format {"mender"};
+
+const ErrorCategoryClass ErrorCategory {};
+
+const char *ErrorCategoryClass::name() const noexcept {
+	return "VersionParserErrorCategory";
+}
+
+string ErrorCategoryClass::message(int code) const {
+	switch (code) {
+	case NoError:
+		return "Success";
+	ParseError:
+		return "Parse error";
+	VersionError:
+		return "Wrong Artifact version";
+	FormatError:
+		return "Wrong Artifact format";
+	default:
+		return "Unknown";
+	}
+}
+
+error::Error MakeError(ErrorCode code, const string &msg) {
+	return error::Error(error_condition(code, ErrorCategory), msg);
+}
+
+
+expected::Expected<Version, error::Error> Parse(io::Reader &reader) {
+	// Read in all the bytes into a tmp buffer
+	std::vector<uint8_t> buf(1024);
+	io::ByteWriter bw {buf};
+
+	auto err = io::Copy(bw, reader);
+	if (err) {
+		return err;
+	}
+
+	auto expected_json = json::LoadFromString(mender::common::StringFromByteVector(buf));
+
+	if (!expected_json) {
+		return MakeError(ParseError, "Failed to parse the version JSON");
+	}
+
+	const json::Json version_json = expected_json.value();
+
+	const int version = version_json.Get("version")
+							.and_then([](const json::Json &j) { return j.GetInt(); })
+							.value_or(-1);
```

I'd drop the `value_or()` and got the `ExpectedInt()` from the chain because `value_or()` discards the error.

---

`value_or()` is still there.

---

Same as above. You need to sleep in more :laughing: